fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/PruneSubfieldsForComplexType.java

```diff
@@ -0,0 +1,588 @@
+            markForNormalOperator(optExpression, context);
+            return null;
+        }
+
+        @Override
+        public Void visitPhysicalSchemaScan(OptExpression optExpression, Context context) {
+            return null;
+        }
+
+        @Override
+        public Void visitPhysicalMysqlScan(OptExpression optExpression, Context context) {
+            return null;
+        }
+
+        @Override
+        public Void visitPhysicalEsScan(OptExpression optExpression, Context context) {
+            return null;
+        }
+
+        @Override
+        public Void visitPhysicalMetaScan(OptExpression optExpression, Context context) {
+            return null;
+        }
+
+        @Override
+        public Void visitPhysicalJDBCScan(OptExpression optExpression, Context context) {
+            return null;
+        }
+
+        @Override
+        public Void visitPhysicalProject(OptExpression optExpression, Context context) {
+            PhysicalProjectOperator physicalProjectOperator = (PhysicalProjectOperator) optExpression.getOp();
+            for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : physicalProjectOperator.getColumnRefMap()
+                    .entrySet()) {
+                entry.getValue().accept(MARK_SUBFIELDS_VISITOR, context);
+            }
+            for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : physicalProjectOperator.getCommonSubOperatorMap()
+                    .entrySet()) {
+                entry.getValue().accept(MARK_SUBFIELDS_VISITOR, context);
+            }
+            return visitChild(optExpression, context);
+        }
+
+        @Override
+        public Void visitPhysicalHashAggregate(OptExpression optExpression, Context context) {
+            markForNormalOperator(optExpression, context);
+            PhysicalHashAggregateOperator physicalHashAggregateOperator =
+                    (PhysicalHashAggregateOperator) optExpression.getOp();
+            if (physicalHashAggregateOperator.getAggregations() != null) {
+                for (Map.Entry<ColumnRefOperator, CallOperator> entry : physicalHashAggregateOperator.getAggregations()
+                        .entrySet()) {
+                    entry.getValue().accept(MARK_SUBFIELDS_VISITOR, context);
+                }
+            }
+            return visitChild(optExpression, context);
+        }
+
+        @Override
+        public Void visitPhysicalTopN(OptExpression optExpression, Context context) {
+            markForNormalOperator(optExpression, context);
+            return visitChild(optExpression, context);
+        }
+
+        @Override
+        public Void visitPhysicalDistribution(OptExpression optExpression, Context context) {
+            markForNormalOperator(optExpression, context);
+            return visitChild(optExpression, context);
+        }
+
+        @Override
+        public Void visitPhysicalHashJoin(OptExpression optExpression, Context context) {
```

Comment:
You already use visitPhysicalJoin, there is no need to write visitPhysicalMergeJoin and visitPhysicalHashJoin